In [1]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [2]:
# Add the project directory to the path
%cd /content/drive/Shared drives/CS634/CS634Project/Q3
import os, sys
sys.path.append(os.getcwd())

/content/drive/Shared drives/CS634/CS634Project/Q3


In [3]:
#Importing necessary library 
import sklearn
import sklearn.ensemble
import sklearn.metrics
import pandas as pd
import numpy as np

In [4]:
data = pd.read_csv("Youtube02-KatyPerry.csv")
data.head()

COMMENT_ID  ... CLASS
0      z12pgdhovmrktzm3i23es5d5junftft3f  ...     1
1    z13yx345uxepetggz04ci5rjcxeohzlrtf4  ...     1
2  z12lsjvi3wa5x1vwh04cibeaqnzrevxajw00k  ...     1
3    z13jcjuovxbwfr0ge04cev2ipsjdfdurwck  ...     1
4  z13qybua2yfydzxzj04cgfpqdt2syfx53ms0k  ...     1

[5 rows x 5 columns]

In [5]:
data.tail()

COMMENT_ID  ... CLASS
345    z12sjp3zgtqnvlysj23zuxxaolrvd1oj504  ...     0
346  z132enrpoy35yxpoe04cjr4zur3jvbyq3xo0k  ...     0
347      z132jbmxfqm4fjysg23nwjfb2mv2vxnua  ...     1
348    z12cdlswetvnejcri04cex0jfwy2u3tzj54  ...     0
349  z120e5uautvcuper304ccf4bjrjugdpbwrc0k  ...     0

[5 rows x 5 columns]

In [6]:
#Checking for null values
data.isnull().sum() 

COMMENT_ID    0
AUTHOR        0
DATE          0
CONTENT       0
CLASS         0
dtype: int64

In [7]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 350 entries, 0 to 349
Data columns (total 5 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   COMMENT_ID  350 non-null    object
 1   AUTHOR      350 non-null    object
 2   DATE        350 non-null    object
 3   CONTENT     350 non-null    object
 4   CLASS       350 non-null    int64 
dtypes: int64(1), object(4)
memory usage: 13.8+ KB


In [8]:
#we are using CONTENT as a independent feature
#the CLASS as dependent label
X = data['CONTENT']
y = data['CLASS']

In [9]:
#Displaying first five records of the feature "CONTENT"
X.head()

0    i love this so much. AND also I Generate Free ...
1    http://www.billboard.com/articles/columns/pop-...
2    Hey guys! Please join me in my fight to help a...
3    http://psnboss.com/?ref=2tGgp3pV6L this is the...
4    Hey everyone. Watch this trailer!!!!!!!!  http...
Name: CONTENT, dtype: object

In [10]:
#Displaying first five records of the label "CLASS"
y.head()

0    1
1    1
2    1
3    1
4    1
Name: CLASS, dtype: int64

**Splitting our dataset into train and test dataset**

In [11]:
#split our dataset into train and test and setting the random state at "123"
#Random state ensures that the splits that you generate are reproducible. Scikit-learn uses random permutations to generate the splits.
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state = 123)
print(X_train.shape)
print(X_test.shape)
print(y_train.shape)
print(y_test.shape)

(280,)
(70,)
(280,)
(70,)


**Vectorizing the dataset**

We use CountVectorizer to convert each content into a matrix of token counts. Moreover, stopwords are removed.

In [12]:
from sklearn.feature_extraction.text import CountVectorizer
vect = CountVectorizer(binary = True, stop_words='english')

In [13]:
## Fitting the CountVectorizer using the training data
vect.fit(X_train)

CountVectorizer(analyzer='word', binary=True, decode_error='strict',
                dtype=<class 'numpy.int64'>, encoding='utf-8', input='content',
                lowercase=True, max_df=1.0, max_features=None, min_df=1,
                ngram_range=(1, 1), preprocessor=None, stop_words='english',
                strip_accents=None, token_pattern='(?u)\\b\\w\\w+\\b',
                tokenizer=None, vocabulary=None)

In [14]:
# Transforming the data frames
X_test_dtm = vect.transform(X_test)
X_train_dtm = vect.fit_transform(X_train)
print(type(X_train_dtm))
print(type(X_test_dtm))

<class 'scipy.sparse.csr.csr_matrix'>
<class 'scipy.sparse.csr.csr_matrix'>


In [15]:
print("Test vector shape : ", X_test_dtm.shape)
print("Train vector shape : ", X_train_dtm.shape)

Test vector shape :  (70, 1278)
Train vector shape :  (280, 1278)


**Logistic Regression Classification model for the Kattyperry data set**

In [16]:
#Importing Logistic regression from Sklearn
from sklearn.linear_model import LogisticRegression
model = LogisticRegression(random_state = 123)
#Fitting the model
model.fit(X_train_dtm,y_train)

LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
                   intercept_scaling=1, l1_ratio=None, max_iter=100,
                   multi_class='auto', n_jobs=None, penalty='l2',
                   random_state=123, solver='lbfgs', tol=0.0001, verbose=0,
                   warm_start=False)

**PREDICTION**

In [17]:
#predicting Label for test data
pred = model.predict(X_test_dtm)
#Checking F1 Score
sklearn.metrics.f1_score(y_test, pred, average='binary')

0.9180327868852458

In [18]:
# Importing necessary packages for metric from Sklearn
from sklearn.metrics import accuracy_score,confusion_matrix,classification_report

In [19]:
Logistic_Accuracy = accuracy_score(y_test,pred)
print('Logistic Regression Accuracy Score: ', Logistic_Accuracy) 

Logistic Regression Accuracy Score:  0.9285714285714286


In [20]:
print('Classification report:\n ' ,classification_report(y_test,pred))

Classification report:
                precision    recall  f1-score   support

           0       0.88      1.00      0.94        37
           1       1.00      0.85      0.92        33

    accuracy                           0.93        70
   macro avg       0.94      0.92      0.93        70
weighted avg       0.94      0.93      0.93        70



**Support Vector Machine(SVM) Classification model (SVC) for the Kattyperry data set**

In [21]:
from sklearn import svm
model_s = svm.SVC(random_state = 123)
model_s.fit(X_train_dtm,y_train)

SVC(C=1.0, break_ties=False, cache_size=200, class_weight=None, coef0=0.0,
    decision_function_shape='ovr', degree=3, gamma='scale', kernel='rbf',
    max_iter=-1, probability=False, random_state=123, shrinking=True, tol=0.001,
    verbose=False)

In [22]:
pred_s = model_s.predict(X_test_dtm)
print(pred_s)

[1 0 0 1 0 0 0 1 1 0 0 0 0 1 1 1 1 0 0 1 0 1 0 0 0 0 1 0 0 1 0 0 0 0 0 1 0
 1 0 0 1 1 1 0 1 1 1 1 0 1 0 1 0 0 1 0 0 0 0 1 0 0 1 1 0 1 0 0 1 0]


In [23]:
SVM_accuracy = accuracy_score(y_test,pred_s)
print('SVM Accuracy Score:', SVM_accuracy)

SVM Accuracy Score: 0.9428571428571428


In [24]:
print(confusion_matrix(y_test, pred_s))
print(classification_report(y_test, pred_s))

[[37  0]
 [ 4 29]]
              precision    recall  f1-score   support

           0       0.90      1.00      0.95        37
           1       1.00      0.88      0.94        33

    accuracy                           0.94        70
   macro avg       0.95      0.94      0.94        70
weighted avg       0.95      0.94      0.94        70



**Random Forest Classification model for the Katty Perry dataset**

In [25]:
#Assigning rf as random forest classifier
rf = sklearn.ensemble.RandomForestClassifier(n_estimators=500, random_state = 125)
#Fiiting the model
rf.fit(X_train_dtm, y_train)

RandomForestClassifier(bootstrap=True, ccp_alpha=0.0, class_weight=None,
                       criterion='gini', max_depth=None, max_features='auto',
                       max_leaf_nodes=None, max_samples=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, n_estimators=500,
                       n_jobs=None, oob_score=False, random_state=125,
                       verbose=0, warm_start=False)

In [26]:
pred_r=rf.predict(X_test_dtm)

In [27]:
RFA = accuracy_score(y_test,pred_r)
print('Random Forest Accuracy Score:', RFA)

Random Forest Accuracy Score: 0.9142857142857143


In [28]:
sklearn.metrics.f1_score(y_test, pred_r, average='binary')

0.9

In [29]:
print(confusion_matrix(y_test,pred_r))
print(classification_report(y_test,pred_r))

[[37  0]
 [ 6 27]]
              precision    recall  f1-score   support

           0       0.86      1.00      0.92        37
           1       1.00      0.82      0.90        33

    accuracy                           0.91        70
   macro avg       0.93      0.91      0.91        70
weighted avg       0.93      0.91      0.91        70



**Naive Bayes Classification model for the Katty Perry data set**

In [30]:
from sklearn.naive_bayes import MultinomialNB
nb = MultinomialNB()

nb.fit(X_train_dtm, y_train) #fit training data

MultinomialNB(alpha=1.0, class_prior=None, fit_prior=True)

In [31]:
y_pred_nb = nb.predict(X_test_dtm) #Run the Prediction

Score the prediction

In [32]:
from sklearn.metrics import accuracy_score, confusion_matrix, roc_auc_score
NBA = accuracy_score(y_test, y_pred_nb)
print('Naive Bayes Accuracy Score:', NBA)

Naive Bayes Accuracy Score: 0.8857142857142857


In [33]:
confusion_matrix(y_test, y_pred_nb)

array([[32,  5],
       [ 3, 30]])

In [34]:
X_test[y_pred_nb > y_test] # False positive comment (That is, not_spam marked as spam)

75                   ROAAAAARRRRRR 🐯🐯🐯
235                                ;-)
167    most popular video on youtube  
171                       Fantastic!!!
263     She kinda let herself go, huh?
Name: CONTENT, dtype: object

In [35]:
X_test[y_pred_nb < y_test] # False negative comment (Spam comment not detected)

247    :: ATTENTION : WE NEED 10 Million Views More f...
54     YAY IM THE 11TH COMMENTER!!!!!                ...
347    KATY PERRY, I AM THE "DÉCIO CABELO", "DECIO HA...
Name: CONTENT, dtype: object

In [36]:
roc_auc_score(y_test, nb.predict_proba(X_test_dtm)[:, 1])

0.9696969696969697

**Analysis**

In [37]:
X_train_tokens = vect.get_feature_names()

In [38]:
nb.feature_count_ # First line represent not_spam, second line represent span

array([[1., 1., 1., ..., 0., 0., 0.],
       [0., 1., 0., ..., 2., 1., 1.]])

In [39]:
not_spam_token_count = nb.feature_count_[0, :]
not_spam_token_count

array([1., 1., 1., ..., 0., 0., 0.])

In [40]:
spam_token_count = nb.feature_count_[1, :]
spam_token_count

array([0., 1., 0., ..., 2., 1., 1.])

In [41]:
tokens = pd.DataFrame({'token': X_train_tokens, 'not_spam': not_spam_token_count, 'spam': 
    spam_token_count}).set_index('token')

# Here are 10 random values drawn from the data frame
# Note that these are absolute occurrences
tokens.sample(10, random_state=50)

not_spam  spam
token                   
coming         0.0   1.0
dede           1.0   0.0
bgq            0.0   1.0
younique       0.0   1.0
opinion        2.0   0.0
link           0.0   3.0
15             1.0   0.0
firework       1.0   0.0
store          0.0   4.0
everyday       0.0   1.0

In [42]:
class_count = nb.class_count_

class_count

array([138., 142.])

In [43]:
tokens.not_spam /= class_count[0]
tokens.spam /= class_count[1]

tokens.sample(10, random_state=50)

not_spam      spam
token                       
coming    0.000000  0.007042
dede      0.007246  0.000000
bgq       0.000000  0.007042
younique  0.000000  0.007042
opinion   0.014493  0.000000
link      0.000000  0.021127
15        0.007246  0.000000
firework  0.007246  0.000000
store     0.000000  0.028169
everyday  0.000000  0.007042

In [44]:
words_with_spam_score = ((tokens.spam + 1e-6)/(tokens.not_spam + 1e-6)).sort_values(ascending=False)

In [45]:
# 50 spam words
words_with_spam_score[:50].keys()

Index(['com', 'www', 'https', 'facebook', 'amp', 'guys', 'help', 'new', 'ref',
       'follow', 'vote', 'visit', 'id', 'cover', 'hi', '2014', 'php', 'click',
       'pages', 'money', 'gofundme', 'shhort', 'hope', 'share', 'apps',
       'thumbs', 'itunes', 'page', 'play', 'support', 'instagram', 'type',
       'soundcloud', 'google', 'store', 'rapper', 'soon', 'single', 'live',
       'link', 'trying', 'water', 'join', 'small', 'covers', 'sub', 'xoxo',
       'getting', 'details', 'download'],
      dtype='object', name='token')

In [46]:
# 50 words which are not spam
words_with_spam_score[-50:].keys()

Index(['clear', 'explicit', 'reply', 'extremely', 'rocks', 'cleanse',
       'empowering', 'champion', 'clip', 'stop', 'millions', 'lets', 'woman',
       'xd', 'soundsofsunday', 'lion', 'wouldn', 'makes', 'didn', 'kitty',
       'seriously', 'fit', 'radio', 'fucking', 'tits', 'purry', 'garbage',
       'favorite', 'acting', 'hating', 'honest', '600', 'doesn', 'illuminati',
       'sexy', 'cute', 'sing', 'daughter', 'billion', 'omg', 'opinion',
       'imagine', 'trash', 'thing', 'official', 'stupid', 'eyes', 'million',
       'looks', 'jungle'],
      dtype='object', name='token')

In [47]:
print('Logistic Regression Accuracy Score: ', Logistic_Accuracy)
print('SVM Accuracy Score:', SVM_accuracy)
print('Random Forest Accuracy Score:', RFA)
print('Naive Bayes Accuracy Score:', NBA)

Logistic Regression Accuracy Score:  0.9285714285714286
SVM Accuracy Score: 0.9428571428571428
Random Forest Accuracy Score: 0.9142857142857143
Naive Bayes Accuracy Score: 0.8857142857142857


**Summary**

In this notebook, we built a machine learning model based on comments for a YouTube video. The training data set had a total of 350 observations. We used a Logistic Regression, SVM, Random Forest and Naive Bayes Classifier in our algorithm, and trained it on 80% of the data set. The model resulted in an accuracy score of **88.57%** for Naive Bayes Classifier, **91.4%** for Random Forest,       **94.28%** for SVM and **92.85%** for Logistic Regression on the test data. We used the model to derive the top 50 most spam and least spam keywords.